mlrun/datastore/base.py

```diff
@@ -36,6 +39,141 @@
 from .utils import filter_df_start_end_time, select_columns_from_df
 
 
+def _get_pyarrow_version():
+    """Get the current PyArrow version."""
+    return version.parse(pyarrow.__version__)
+
+
+def is_pyarrow_17_or_later():
+    """Check if PyArrow version is 17.0.0 or later."""
+    return _get_pyarrow_version() >= version.parse("17.0.0")
+
+
+def _detect_partition_data_conflict(error_message):
+    """
+    Detect if the error is a PyArrow 17+ schema conflict between partition and data columns.
+
+    Args:
+        error_message (str): The error message from PyArrow
+
+    Returns:
+        dict: Contains conflict information or None if not a conflict
+    """
+    # Only check for conflicts in PyArrow 17+
+    if not is_pyarrow_17_or_later():
+        return None
+
+    # Pattern to match PyArrow schema conflict errors
+    conflict_pattern = (
+        r"Unable to merge: Field (\w+) has incompatible types: (.+?) vs (.+)$"
+    )
+
+    match = re.search(conflict_pattern, error_message.strip())
+    if not match:
+        return None
+
+    field_name = match[1]
+    type1 = match[2].strip()
+    type2 = match[3].strip()
+
+    # Check if this looks like a partition vs data column conflict
+    # Partition columns are typically string type
+    is_partition_conflict = (
+        (type1 == "string" and "dictionary" in type2)
+        or (type2 == "string" and "dictionary" in type1)
+        or (type1 == "string" and type2 != "string")
+        or (type2 == "string" and type1 != "string")
+    )
+
+    if not is_partition_conflict:
+        return None
```

Comment:
I don't really see what this has to do with dictionary types? If I were to make an assumption regarding the original type, it would be int (e.g. year, month), not dict.